dnsnameresolver.go

```diff
@@ -0,0 +1,289 @@
+			if !resolver.configuredNamespace(resolverObj.Namespace) {
+				return
+			}
+
+			dnsName := string(resolverObj.Spec.Name)
+			// Check if the DNS name is wildcard or regular.
+			if isWildcard(dnsName) {
+				// If the DNS name is wildcard, delete the details of the DNSNameResolver
+				// object from the wildcardDNSInfo map.
+				resolver.wildcardMapLock.Lock()
+				if dnsInfoMap, exists := resolver.wildcardDNSInfo[dnsName]; exists {
+					// If details of DNS name and the DNSNameResolver objects already exist
+					// then check if the existing information match with the current one.
+					// Otherwise, don't proceed.
+					if dnsInfoMap[resolverObj.Namespace] == resolverObj.Name {
+						delete(dnsInfoMap, resolverObj.Namespace)
+						if len(dnsInfoMap) > 0 {
+							resolver.wildcardDNSInfo[dnsName] = dnsInfoMap
+						} else {
+							delete(resolver.wildcardDNSInfo, dnsName)
+						}
+					}
+				}
+				resolver.wildcardMapLock.Unlock()
+			} else {
+				// If the DNS name is regular, delete the details of the DNSNameResolver
+				// object from the regularDNSInfo map.
+				resolver.regularMapLock.Lock()
+				if dnsInfoMap, exists := resolver.regularDNSInfo[dnsName]; exists {
+					// If details of DNS name and the DNSNameResolver objects already exist
+					// then check if the existing information match with the current one.
+					// Otherwise, don't proceed.
+					if dnsInfoMap[resolverObj.Namespace] == resolverObj.Name {
+						delete(dnsInfoMap, resolverObj.Namespace)
+						if len(dnsInfoMap) > 0 {
+							resolver.regularDNSInfo[dnsName] = dnsInfoMap
+						} else {
+							delete(resolver.regularDNSInfo, dnsName)
+						}
+					}
+				}
+				resolver.regularMapLock.Unlock()
+			}
+
+			// Used only in unit tests.
+			if send != nil {
+				send(resolverObj)
+			}
+		},
+		// Used only in unit tests.
+		// Update event.
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			// Get the DNSNameResolver object.
+			newResolverObj, ok := oldObj.(*ocpnetworkapiv1alpha1.DNSNameResolver)
+			if !ok {
+				log.Infof("object not of type DNSNameResolver: %v", oldObj)
+				return
+			}
+
+			// Check if namespace is configured or not.
+			if !resolver.configuredNamespace(newResolverObj.Namespace) {
+				return
+			}
+
+			// Used only in unit tests.
+			if send != nil {
+				send(newResolverObj)
+			}
+		},
+	})
+	return nil
+}
+
+// initPlugin initializes the ocp_dnsnameresolver plugin and returns the plugin startup and
+// shutdown callback functions.
+func (resolver *OCPDNSNameResolver) initPlugin() (func() error, func() error, error) {
+	// Create a client supporting network.openshift.io apis.
+	kubeConfig, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	networkClient, err := ocpnetworkclient.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = resolver.initInformer(networkClient, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resolver.stopCh = make(chan struct{})
```

Comment:
`stopCh` doesn't need to be a field of `OCPDNSNameResolver`, it's used only in `initPlugin` function.

Reply:
Another thing which just popped in my mind. Normally a stop channel is either a timeout (ticker) or a signal handler. Here we create a stop channel just because the informer's `Run()` method needs one. So basically when the CoreDNS will be stopped (by a signal as we are in a Kubernetes environment) this bit of information won't be passed down to our plugin's informer.

Is there a way to handle external signals inside CoreDNS plugins?